pkg/ucp/store/inmemory/client.go

```diff
@@ -0,0 +1,262 @@
+// This is provided for ease of debugging.
+//
+// Here's an example:
+//
+//	resource ID: "/planes/radius/local/resourceGroups/my-rg/providers/Applications.Test/testType1/my-resource/testType2/my-child-resource"
+//
+// The entry would be:
+//
+//	rootScope: "/planes/radius/local/resourceGroups/my-rg"
+//	resourceType: "Applications.Test/testType1/testType2"
+//	routingScope: "Applications.Test/testType1/my-resource/testType2/my-child-resource"
+//
+// All fields are compared case-insensitively.
+type entry struct {
+	obj store.Object
+
+	// rootScope is the root scope of the resource ID.
+	rootScope string
+
+	// resourceType is the resource type of the resource ID.
+	resourceType string
+
+	// routingScope is the routing scope of the resource ID.
+	routingScope string
+}
+
+// NewClient creates a new in-memory store client.
+func NewClient() *Client {
+	return &Client{
+		mutex:     sync.Mutex{},
+		resources: map[string]entry{},
+	}
+}
+
+// Get implements store.StorageClient.
+func (c *Client) Get(ctx context.Context, id string, options ...store.GetOptions) (*store.Object, error) {
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	entry, ok := c.resources[strings.ToLower(normalized.String())]
+	if !ok {
+		return nil, &store.ErrNotFound{ID: id}
+	}
+
+	return &entry.obj, nil
+}
+
+// Delete implements store.StorageClient.
+func (c *Client) Delete(ctx context.Context, id string, options ...store.DeleteOptions) error {
+	if ctx == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	config := store.NewDeleteConfig(options...)
+
+	entry, ok := c.resources[strings.ToLower(normalized.String())]
+	if !ok && config.ETag != "" {
```

Comment:
Could you explain this? If entry is not found and `config.ETag` is not empty, why is it a Concurrency error? If entry is found and ETag is not the most recent then I think that should be a concurrency issue.

Reply:
First, this isn't new behavior. The tests for the data store have been around for a long time. Second, I think it's the least surprising behavior and therefore the most correct thing to do.

Any operation that uses ETags is an example of *optimistic concurrency control*. Here's a relevant snippet from the [wikipedia description](https://en.wikipedia.org/wiki/Optimistic_concurrency_control).

> While running, transactions use data resources without acquiring locks on those resources. Before committing, each transaction verifies that no other transaction has modified the data it has read. If the check reveals conflicting modifications, the committing transaction rolls back and can be restarted.

This supports the following pattern:

- An actor reads the current state of the resource (including it's ETag).
- It applies some business logic based on the state of the resource and decides to delete the resource.
- It uses the ETag so that the delete operation will **ONLY** be persisted if the resource is unchanged.

There are three possible cases:

- The resource was changed (and has a new ETag).
- The resource was deleted.
- The resource is unchanged and will be deleted.

The behavior of the datastore collapses the first two cases into a single one.

Keep in mind that most code will not use ETags.